[PowerPC] Support conversion between f16 and f128

EsmeYi opened the pull request:
This patch enables conversion between f16 and f128, expanding on pre-Power9 targets and using HW instructions on Power9.
Fixes https://github.com/llvm/llvm-project/issues/92866

Commit:
F16 and f128 conversion.

tgross35 commented:
Should tests at [`f128-conv.ll`](https://github.com/llvm/llvm-project/blob/ad82d1c53f089937c05af11ff45798ceb5ca894e/llvm/test/CodeGen/PowerPC/f128-conv.ll) or [`fp128-libcalls.ll`](https://github.com/llvm/llvm-project/blob/ad82d1c53f089937c05af11ff45798ceb5ca894e/llvm/test/CodeGen/PowerPC/fp128-libcalls.ll) be updated?

efriedma-quic reviewed (changes requested):
For truncation, if you convert f128->f64->f16, you'll round twice, and therefore get the wrong result in some cases.  You need to ensure you only round once.

Commit:
For truncation from f128 to f16, always use the lib fucntion (direct conversion) instead of HW instructions (stepwise conversion).

EsmeYi commented:
> For truncation, if you convert f128->f64->f16, you'll round twice, and therefore get the wrong result in some cases. You need to ensure you only round once.

Thanks Eli @efriedma-quic ! This is a great find. I just realized that the intermediate rounding during the stepwise conversion (f128->f64->f16) can produce a different result compared to the direct conversion (`f128->f16).

Commit:
Updated tests.

EsmeYi commented:
> Should tests at [`f128-conv.ll`](https://github.com/llvm/llvm-project/blob/ad82d1c53f089937c05af11ff45798ceb5ca894e/llvm/test/CodeGen/PowerPC/f128-conv.ll) or [`fp128-libcalls.ll`](https://github.com/llvm/llvm-project/blob/ad82d1c53f089937c05af11ff45798ceb5ca894e/llvm/test/CodeGen/PowerPC/fp128-libcalls.ll) be updated?

Thank you @tgross35. Sorry that I missed the comment before. I've updated the 2 tests and removed llvm/test/CodeGen/PowerPC/f16-to-from-f128.ll, since it was redundant then.

EsmeYi commented:
Gentle ping.

chenzheng1030 commented on llvm/test/CodeGen/PowerPC/f128-conv.ll:
```diff
@@ -1450,6 +1450,70 @@ entry:
+  ret half %0
+}
+
+define fp128 @ext(half %a) unnamed_addr {
```
nit: Is the `unnamed_addr` needed?

chenzheng1030 commented on llvm/test/CodeGen/PowerPC/f128-conv.ll:
```diff
@@ -1450,6 +1450,70 @@ entry:
+; CHECK-P8-NEXT:    mtlr r0
+; CHECK-P8-NEXT:    blr
+entry:
+  %0 = fptrunc fp128 %a to half
```
nit: We'd better avoid the unnamed variable names, like `%0`. (Suggested by https://llvm.org/docs/TestingGuide.html#writing-new-regression-tests)

chenzheng1030 commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -210,13 +210,19 @@ PPCTargetLowering::PPCTargetLowering(const PPCTargetMachine &TM,
     setLoadExtAction(ISD::SEXTLOAD, VT, MVT::i8, Expand);
   }
 
+  setTruncStoreAction(MVT::f128, MVT::f16, Expand);
```
nit: could you please add a comment here why we can not use PWR9 instructions to do the conversion.

chenzheng1030 reviewed (approved):
Looks reasonable to me except some nits which can be addressed in the merge commit. Thanks for fixing this.

tgross35 commented:
@EsmeYi are you able to get this over the line? Looks like it just needs a few small tweaks.

lei137 commented:
Commandeering this patch to get this feature in.
https://github.com/llvm/llvm-project/pull/130158
